File: Target LeetdCode 200/Matchsticks_to_Square.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
   public:
    bool solve(vector<int>& nums, int w, int x, int y, int z, int i) {
        if (i == nums.size()) {
            return (x == y && y == z && x == w) && i > 3;
        }
        return solve(nums, w + nums[i], x, y, z, i + 1) ||
               solve(nums, w, x + nums[i], y, z, i + 1) ||
               solve(nums, w, x, y + nums[i], z, i + 1) ||
               solve(nums, w, x, y, z + nums[i], i + 1);
    }
    bool makesquare(vector<int>& nums) {
        return solve(nums, 0, 0, 0, 0, 0);
    }
};

int main() {
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
}

File: CSES-KickStart-CF/MatchsticksToSquare.java
class Solution {
    int side[];
    private boolean allEqual() {
        for (int i = 0; i < 3; i++) {
            if (side[i] != side[i + 1] || side[i] == 0 || side[i + 1] == 0) return false;
        }
        return true;
    }

    private boolean solve(int [] nums, int i) {
        if (i == nums.length) {
            return allEqual();
        }

        boolean ans = false;

        for (int j = 0; j < 4 && !ans; j++) {
            side[j] += nums[i];
            ans |= solve(nums, i + 1);
            side[j] -= nums[i];
        }
        return ans;
    }

    public boolean makesquare(int[] stick) {
        side = new int[4];
        Arrays.fill(side, 0);
        return solve(stick, 0);
    }
}